hack/generator/pkg/astmodel/storage_conversion_endpoint.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"github.com/gobuffalo/flect"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+type KnownLocalsSet map[string]struct{}
+
+// StorageConversionEndpoint represents either a source or a destination field for a storage conversion
+type StorageConversionEndpoint struct {
+	// theType is the Type of the value accessible via this endpoint
+	theType Type
+	// name is the name of the underlying property, used to generate useful local identifiers
+	name string
+	// knownLocals is a shared map of locals that have already been created within a given function, to prevent duplicates
+	knownLocals KnownLocalsSet
+}
+
+func NewStorageConversionEndpoint(theType Type, name string, knownLocals KnownLocalsSet) *StorageConversionEndpoint {
+	return &StorageConversionEndpoint{
+		theType:     theType,
+		name:        strings.ToLower(name),
+		knownLocals: knownLocals,
+	}
+}
+
+// Expr returns a clone of the expression for this endpoint
+// (returning a clone avoids issues with reuse of fragments within the dst)
+func (endpoint *StorageConversionEndpoint) Type() Type {
+	return endpoint.theType
+}
+
+// Create an identifier for a local variable that represents a single item
+// Each call will return a unique identifier
+func (endpoint *StorageConversionEndpoint) CreateSingularLocal() string {
+	singular := flect.Singularize(endpoint.name)
+	return endpoint.knownLocals.createLocal(singular)
+}
+
+// CreatePluralLocal creates an identifier for a local variable that represents multiple items
+// Each call will return a unique identifier
+func (endpoint *StorageConversionEndpoint) CreatePluralLocal(suffix string) string {
+	plural := flect.Pluralize(endpoint.name)
+	return endpoint.knownLocals.createLocal(plural + suffix)
+}
+
+// WithType creates a new endpoint with a different type
+func (endpoint *StorageConversionEndpoint) WithType(theType Type) *StorageConversionEndpoint {
+	return NewStorageConversionEndpoint(
+		theType,
+		endpoint.name,
+		endpoint.knownLocals)
+}
+
+// createLocal creates a new unique local with the specified suffix
+// Has to be deterministic, so we use an incrementing number to make them unique
+func (locals KnownLocalsSet) createLocal(nameHint string) string {
```

Comment:
It feels like there's some duplication or overlap here and with `IdentifierFactory`?

Obviously `IdentifierFactory` doesn't have any of the locals-tracking stuff, but wondering if it makes sense to do some or all of the following:
1. Promote this concept of the `KnownLocalsSet` to its own file (possibly throwing in a rename to use `Identifier` somewhere in there? `KnownLocalIdentifierSet`?)
2. Take `IdentifierFactory` as an input to a `NewSet` constructor.
3. Move any functionality missing from `IdentifierFactory` there (the plural/singular bits?)

That has the added advantage (at least I think it's an advantage?) that the renames that we set up in `idFactory` apply to the locals as well? Although I guess maybe they already do because the namehint is coming from something which has already been renamed by the idFactory? Also allows this to make use of @porges caching stuff as well.

Reply:
Definitely at least the `toPrivate` bit of this is exactly what `IdentifierFactory` `CreateIdentifier` with visibility `notexported` already does.

Reply:
I promoted `KnownLocalsSet` from an alias to a full type, added an `IdentifierFactory` as a member and created a factory method `NewKnownLocalsSet()`. Nice and tidy.

Reply:
Cool, I'll make a note when this merges to go see if I can find what I did in the ARM functions and maybe update to use `KnownLocalsSet`, as there's certainly a similar problem there and we could probably benefit from reuse.